src/node_url.cc

```diff
@@ -1545,44 +1546,61 @@ void URL::Parse(const char* input,
 }  // NOLINT(readability/fn_size)
 
 // https://url.spec.whatwg.org/#url-serializing
-std::string URL::SerializeURL(const struct url_data* url,
+std::string URL::SerializeURL(const url_data& url,
                               bool exclude = false) {
-  std::string output = url->scheme;
-  if (url->flags & URL_FLAGS_HAS_HOST) {
+  std::string output;
+  output.reserve(
+    10 +
+    url.scheme.size() +
+    url.username.size() +
+    url.password.size() +
+    url.host.size() +
+    url.query.size() +
+    url.fragment.size() +
+    url.href.size() +
+    std::accumulate(
+        url.path.begin(),
+        url.path.end(),
+        0,
+        [](size_t sum, const auto& str) { return sum + str.size(); }));
+
+  output += url.scheme;
+  if (url.flags & URL_FLAGS_HAS_HOST) {
     output += "//";
-    if (url->flags & URL_FLAGS_HAS_USERNAME ||
-        url->flags & URL_FLAGS_HAS_PASSWORD) {
-      if (url->flags & URL_FLAGS_HAS_USERNAME) {
-        output += url->username;
+    if (url.flags & URL_FLAGS_HAS_USERNAME ||
+        url.flags & URL_FLAGS_HAS_PASSWORD) {
+      if (url.flags & URL_FLAGS_HAS_USERNAME) {
+        output += url.username;
       }
-      if (url->flags & URL_FLAGS_HAS_PASSWORD) {
-        output += ":" + url->password;
+      if (url.flags & URL_FLAGS_HAS_PASSWORD) {
+        output += ":" + url.password;
       }
       output += "@";
     }
-    output += url->host;
-    if (url->port != -1) {
-      output += ":" + std::to_string(url->port);
+    output += url.host;
+    if (url.port != -1) {
+      output += ":" + std::to_string(url.port);
     }
   }
-  if (url->flags & URL_FLAGS_CANNOT_BE_BASE) {
-    output += url->path[0];
+  if (url.flags & URL_FLAGS_CANNOT_BE_BASE) {
+    output += url.path[0];
   } else {
-    if (!(url->flags & URL_FLAGS_HAS_HOST) &&
-          url->path.size() > 1 &&
-          url->path[0].empty()) {
+    if (!(url.flags & URL_FLAGS_HAS_HOST) &&
+          url.path.size() > 1 &&
+          url.path[0].empty()) {
       output += "/.";
     }
-    for (size_t i = 1; i < url->path.size(); i++) {
-      output += "/" + url->path[i];
+    for (size_t i = 1; i < url.path.size(); i++) {
+      output += "/" + url.path[i];
     }
   }
-  if (url->flags & URL_FLAGS_HAS_QUERY) {
-    output = "?" + url->query;
+  if (url.flags & URL_FLAGS_HAS_QUERY) {
+    output += "?" + url.query;
   }
-  if (!exclude && url->flags & URL_FLAGS_HAS_FRAGMENT) {
-    output = "#" + url->fragment;
+  if (!exclude && url.flags & URL_FLAGS_HAS_FRAGMENT) {
```

Comment:
```suggestion
  if (!exclude && (url.flags & URL_FLAGS_HAS_FRAGMENT)) {
```

Sorry, I can never remember the operator precedence here (and I suspect many others don't either)